syntax/parser_arithm.go

```diff
@@ -0,0 +1,318 @@
+package syntax
+
+// compact specifies whether we allow spaces between expressions.
+// This is true for let
+func (p *Parser) arithmExpr(compact bool) ArithmExpr {
+	return p.arithmExprComma(compact)
+}
+
+// These function names are inspired by Bash's expr.c
+
+func (p *Parser) arithmExprComma(compact bool) ArithmExpr {
+	return p.arithmExprBinary(compact, p.arithmExprAssign, Comma)
+}
+
+func (p *Parser) arithmExprAssign(compact bool) ArithmExpr {
+	// Assign is different from the other binary operators because it's right-associative and needs to check that it's placed after a name
+	value := p.arithmExprCond(compact)
+	switch BinAritOperator(p.tok) {
+	case AddAssgn, SubAssgn, MulAssgn, QuoAssgn, RemAssgn, AndAssgn,
+		OrAssgn, XorAssgn, ShlAssgn, ShrAssgn, Assgn:
+		if compact && p.spaced {
+			return value
+		}
+		if !isArithName(value) {
+			p.posErr(p.pos, "%s must follow a name", p.tok.String())
+		}
+		pos := p.pos
+		tok := p.tok
+		p.nextArithOp(compact)
+		y := p.arithmExprAssign(compact)
+		if y == nil {
+			p.followErrExp(pos, tok.String())
+		}
+		return &BinaryArithm{
+			OpPos: pos,
+			Op:    BinAritOperator(tok),
+			X:     value,
+			Y:     y,
+		}
+	}
+	return value
+}
+
+func (p *Parser) arithmExprCond(compact bool) ArithmExpr {
```

Comment:
nit: perhaps replace Cond with Tern, since that's probably less ambiguous to the reader